Issue #9: "Resource" Content Type

lrowang opened the pull request:
Hi @lhridley  - 

PR for #9.

I created a test suite for the custom content type "resource" and its fields. I think I'm close, I think there's still a few steps that I'm missing/haven't gotten correctly.

1. The `behat.yml` file. I followed your lead on the `taxonomy.feature` and this [guide](http://behat-drupal-extension.readthedocs.io/en/latest/blackbox.html), and created some "region maps" in the `behat.yml` file for the new `resource.feature`. I used the "inspect" tool in the browser and picked out a CSS ID for what I think is the right region on the page, but I'm not entirely sure that's the right method -- would you mind verifying?

2. For some reason my new test suite is still "undefined." Based on the message when I run `make test`, I suspect I'm supposed to create a `Context.php` file specifically for this test suite? But if that's the case, how come the "taxonomy" test suite doesn't have one?
![casestudy 9-snippetmsg](https://cloud.githubusercontent.com/assets/19154706/18112142/57b7c5ea-6eda-11e6-9243-5f29b9bf0978.png)

3. I think this is related to the above, but I also can't seem to run the test suite independently. GAH!
![casestudy 9-runindependent](https://cloud.githubusercontent.com/assets/19154706/18112161/71f937a4-6eda-11e6-9076-88fbd6f80cad.png)

Thanks,
Ro

lhridley commented on tests/behat.yml:
```diff
@@ -67,6 +67,11 @@ default:
+      paths:
+        - %paths.base%/features/content-types
+      contexts:
+        - Drupal\DrupalExtension\Context\DrupalContext
```
You can add a context here for any custom steps you may need, and 'behat --init' will create it for you.   For your content types, just call it ContentTypeContext.

lhridley commented on tests/features/content-types/contentresource.feature:
```diff
@@ -0,0 +1,23 @@
+  Scenario Outline:  Check the "Resource" content type's fields exist
+    Given I am logged in as a user with the staff role
+    When I visit "/admin/structure/types/manage/resource/fields"
+    Then I should see "<fields>" in the "Resources" region
```
You've done a good job here, but I would probably call this the "Fields" region instead of the "Resources" region -- you can then use it across all your content type tests and it will be more applicable.

lhridley commented on tests/behat.yml:
```diff
@@ -85,3 +91,5 @@ default:
         footer: "#footer"
         content: ".content"
         taxonomy: "#taxonomy"
+        Content types: "#overlay-content"
```
I would look for a selector that is not tied to the "Overlay" module.  This can be turned on and off (I for one don't like it and will frequently disable it), and when it's disabled any "overlay" selectors are no longer available.

You can use the defined region "content" for the step where you're looking for the Resource content type.

lhridley commented on tests/behat.yml:
```diff
@@ -85,3 +91,5 @@ default:
         content: ".content"
         taxonomy: "#taxonomy"
+        Content types: "#overlay-content"
+        Resources: "#field-overview"
```
Be a little more generic here, I would use "Fields" region instead of "Resources" region -- then you can use it on all your content type tests and have it be more readable.

lhridley commented on tests/features/content-types/resource.feature:
```diff
@@ -0,0 +1,23 @@
+
+  @api
+  Scenario:  Check the "Resource" content type exists
+    Given I am logged in as a user with the staff role
```
Users with the "staff" role (you need to put staff in quotations here for the predefined steps to work properly) don't have permission to create or modify content types, nor do we want them to.  This will always fail, but it's a good test for the "administrator" content type.

However your "staff" members will need to be able to create and manage content, so a better test would be to make sure they can add, edit, publish, unpublish, and delete content for each of the content types once the content types are created.

lhridley commented on tests/features/content-types/resource.feature:
```diff
@@ -0,0 +1,23 @@
+
+  @api
+  Scenario Outline:  Check the "Resource" content type's fields exist
+    Given I am logged in as a user with the staff role
```
See above comments for "staff" vs "administrator" role.

lhridley commented:
Ro,

Excellent first stab, and yes your methodology for identifying a CSS selector for your region definition was perfect, that's exactly how you do it.

I've made several comments here, if you have questions about any of them and want to discuss, ping me and we can jump on a call or have a Slack chat.

lrowang commented:
Hi @lhridley - 

Thank you for the helpful comments! I made the relevant changes to the Behat tests and was able to get the test suite running. Here's a summary and lots of questions! 😛 

### Resource
![casestudy 9-fields](https://cloud.githubusercontent.com/assets/19154706/18144103/d93b1618-6f79-11e6-9334-e3b9e004123a.png)

- I created the custom "Resource" content type and it's corresponding fields. I downloaded and enabled the "Links" module in order to create a custom URL link field -- **I think that was the right thing to do, but let me know if I complicated things unnecessarily.** (is `URL path settings` something that came from `PathAuto`?)
- I think I set up the "Visitor Type" and "Resource Category" fields properly. **Is "Unlimited" the right "Number of Values" since it's essentially a tag?** My reasoning is that you should be able to tag all the taxonomy terms if you want to.
- Or instead of creating special "Visitor Type" and "Resource Category" fields, **should I have just used the existing "Tag" field?** (I didn't, but I could see it working this way too).
- I'm not sure how to hide the "visitor" and "resource" types **just for visitors**. I was able to hide them completely by going to `Manage Display` tab, but that ends up hiding those terms for _everybody_. I checked this by creating a "Test Resource" content and then viewing it as `admin` and `anonymous`. 

![casestudy 9-hidden](https://cloud.githubusercontent.com/assets/19154706/18144329/e1fe5af2-6f7a-11e6-9c30-b20b37e4efbf.png)

### Behat Testing
- When I created the new "Resource" content, it generated new permissions that I had to add to the staff and administrator roles. I also added them to the`staff` user's granted permissions in the `behat.yml` file -- **is this enough to test if the `staff` user is able to create/edit/delete the new content?** I ask because I also created/proposed some scenarios in the `resource.feature` file, but they seem redundant?

![casestudy 9-permissions](https://cloud.githubusercontent.com/assets/19154706/18144489/bff12128-6f7b-11e6-8041-f14836beb663.png)

#### Also, all tests passing!
![casestudy 9-tpass](https://cloud.githubusercontent.com/assets/19154706/18144549/0502b6dc-6f7c-11e6-9bbc-c707bd7a4532.png)

lrowang commented:
AHA! fixed the darn builds! Take THAT typo!

lrowang commented:
Hi @lhridley -

This PR is finally ready for official review!

I updated the Behat tests so that they:
- Check if `Resource` content exists and has the right fields.
- Checks if the "staff" user can create/edit/delete `resource content`
- Creates 5 `resource` instances and checks they are displaying the right fields (per the issue instructions)

Also, I fixed the problem of local tests passing but Travis failing -- it was a dumb typo! BAH!

Thanks,
Ro
